### PR TITLE
[.NET 8 in-proc] Multi-target test projects

### PR DESCRIPTION
**Pull request by liliankasem**
### Issue describing the changes in this PR

Part of #10044 

**IMPORTANT**: Currently, changes must be backported to the `in-proc` branch to be included in Core Tools and non-Flex deployments.

* [x] Backporting to the `in-proc` branch is not required
    * Otherwise: Link to backporting PR
* [x] My changes **do not** require documentation changes
    * [ ] Otherwise: Documentation issue linked to PR
* [x] My changes **should not** be added to the release notes for the next release
    * [ ] Otherwise: I've added my notes to `release_notes.md`
* [x] My changes **do not** need to be backported to a previous version
    * [ ] Otherwise: Backport tracked by issue/PR #issue_or_pr
* [x] My changes **do not** require diagnostic events changes
    * Otherwise: I have added/updated all related diagnostic events and their documentation (Documentation issue linked to PR)
* [x] I have added all required tests (Unit tests, E2E tests)

### Additional Information

- Updated test projects to multi-target net6 and net8
- Updates ADO pipeline to run tests for both targets
- Updated unit test to separate out EGM from other unit tests (this is a temporary change whilst @brettsam works on multi-targeting for the EGM project and it's test - consider it out of scope for this PR).

### CODE REVIEW

**Comment by liliankasem** [comment]
@fabiocav - should I target this change into your clean branch for inproc8?

File: test/WebJobs.Script.Tests.Integration/WebJobs.Script.Tests.Integration.csproj
Comment (jviau) [inline test/WebJobs.Script.Tests.Integration/WebJobs.Script.Tests.Integration.csproj]: You could simplify this by doing the following:

``` xml
<MSBuild Projects="..\CSharpPrecompiledTestProjects\CSharpPrecompiledTestProjects.sln"
             Targets="Publish"
             Properties="Configuration=Debug;TargetFramework=$(TargetFramework)'"
             Condition="'$(TargetFramework)' != ''" />
```

How multi-TFM builds work is there is an outer and inner build(s). Outer build just kicks off an inner build per TFM. So the above does two things:

1. Detect we are in an inner build by checking `Condition="'$(TargetFramework)' != ''"` (this _should_ be empty on outer build).
2. Pass TFM to the `Publish` command.

File: test/WebJobs.Script.Tests/WebJobs.Script.Tests.csproj
Comment (jviau) [inline test/WebJobs.Script.Tests/WebJobs.Script.Tests.csproj]: Can simplify this to:

``` XML
<None Update="DepsFiles\$(TargetFramework)\Microsoft.Azure.WebJobs.Script.WebHost.deps.json"
    CopyToOutputDirectory="Always"
    Link="Microsoft.Azure.WebJobs.Script.WebHost.deps.json"
    Condition="'$(TargetFramework)" != ''" />
```

The `Condition="'$(TargetFramework)" != ''" ` may not even be necessary

File: test/CSharpPrecompiledTestProjects/Dependency55/Dependency55.csproj
Comment (jviau) [inline test/CSharpPrecompiledTestProjects/Dependency55/Dependency55.csproj]: Can you explain this change? It shouldn't need to update - is there some issue you are working around here?

File: test/WebJobs.Script.Tests.Integration/WebJobs.Script.Tests.Integration.csproj
Comment (liliankasem) [reply test/WebJobs.Script.Tests.Integration/WebJobs.Script.Tests.Integration.csproj]: Thanks for the explanation, this helps a lot!

File: test/CSharpPrecompiledTestProjects/Dependency55/Dependency55.csproj
Comment (liliankasem) [reply test/CSharpPrecompiledTestProjects/Dependency55/Dependency55.csproj]: When I switched the integration test project to multi-target, I originally got the following error message:

> C:\hostedtoolcache\windows\dotnet\sdk\8.0.204\Sdks\Microsoft.NET.Sdk\targets\Microsoft.PackageDependencyResolution.targets(266,5): error NETSDK1005: Assets file 'D:\a\_work\1\s\out\obj\Dependency56\project.assets.json' doesn't have a target for 'net6.0'. Ensure that restore has run and that you have included 'net6.0' in the TargetFrameworks for your project.

I didn't quite understand why it was just the Dependency 55 and 56 projects that are impacted by this?

I just changed this back to singular (`TargetFramework`) and I'm not seeing this error message anymore. 

(Also if you have any resources/docs/courses you would recommend to learn more about MSBuild and these project files, I would be very interested in learning more)

Comment (liliankasem) [reply test/CSharpPrecompiledTestProjects/Dependency55/Dependency55.csproj]: Actually failing in the pipeline now with the last 2 commits I made :(

https://azfunc.visualstudio.com/Azure%20Functions/_build/results?buildId=165646&view=logs&j=18d94a44-837b-59b4-84e6-015da6d5bdc8&t=2ae8e128-7a8d-54f3-a559-ea229ccd22c0

Comment (liliankasem) [reply test/CSharpPrecompiledTestProjects/Dependency55/Dependency55.csproj]: AHHHHHH; it's an extra `'` in `Properties="Configuration=Debug;TargetFramework=$(TargetFramework)'"` - fixing ☠️

File: test/WebJobs.Script.Tests.Integration/WebJobs.Script.Tests.Integration.csproj
Comment (jviau) [inline test/WebJobs.Script.Tests.Integration/WebJobs.Script.Tests.Integration.csproj]: You can use a glob if you want:

``` xml
    <ProjectReference Include="..\CSharpPrecompiledTestProjects\**\*.csproj" ReferenceOutputAssembly="false" />
```

You also will need to add to the solution, otherwise VS won't build it (VS tries to be all fancy with its build, I don't like it 🤣)

I like to use dotnet cli for that;
From the repo root:

``` pwsh
Get-ChildItem .\test\CSharpPrecompiledTestProjects\ -Recurse -Filter *.csproj | % { dotnet sln WebJobs.Script.sln add $_.FullName }
```

Comment (liliankasem) [reply test/WebJobs.Script.Tests.Integration/WebJobs.Script.Tests.Integration.csproj]: Thank you for this, will update

**Comment by liliankasem** [comment]
@fabiocav pipeline is all green except for the artifact build conflict due to having to rerun for a flakey test - is it okay if we bypass and merge (once we have approvals):

https://azfunc.visualstudio.com/Azure%20Functions/_build/results?buildId=165693&view=results

I wrote up a Teams thread about this pipeline issue for us to follow up on.

**Review by jviau — APPROVED** [review]

**Review by fabiocav — APPROVED** [review]

**Pull request merged**
